### PR TITLE
Mad/metrics: Generation of many R1CS systems for benchmarking proof systems

### PR DESCRIPTION
**Pull request by Michael-Qedit**
Generation of many R1CS systems for benchmarking proof systems

### CODE REVIEW

File: rust/src/producers/circuit_generator.rs [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    for biguint in biguints.iter() {
+        ret.append(&mut serialize_biguint(biguint.clone(), byte_len));
+    }
+    ret
+}
+
+fn compute_equation(witnesses: &Vec<BigUint>, bit_vector: &[u8]) -> BigUint {
+    let mut ret = BigUint::zero();
+    let _: Vec<usize> = witnesses.iter().enumerate()
+                .filter(|(idx, _buint)| bit_vector[*idx] != 0 )
```
Comment (PaulLaux) [inline rust/src/producers/circuit_generator.rs:226]: `_buint` is unused, and the filter is not very intuetive. 
I would go with somthing like
```rust
fn compute_equation(witnesses: &Vec<BigUint>, bit_vector: &Vec<u8>) -> BigUint {
    witnesses.iter().enumerate().map(|(idx, buint)| {
        buint * bit_vector[idx];
    }).sum::<BigUint>();
}
```

**Review by PaulLaux — COMMENTED** [review]
Great work!
Added some comments, mostly about structure.
Also, please provide a description of this PR.

File: rust/src/producers/circuit_generator.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+///
+/// # Examples
+/// To generate a R1CS system with 10 witnesses, 20 instance variables, over the field F_{257}
+/// ```
+///     use std::fs::remove_dir_all;
+///     use std::path::PathBuf;
+///     use zkinterface::producers::circuit_generator::generate_metrics_data;
+///
+///     let workspace = PathBuf::from("local/test_metrics");
+///     let _ = remove_dir_all(&workspace);
```
[the remark this answers is not in the document]
Comment (Michael-Qedit) [reply rust/src/producers/circuit_generator.rs:136]: Done

File: rust/src/producers/circuit_generator.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    let workspace = PathBuf::from("local/test_metrics");
+    let _ = remove_dir_all(&workspace);
+
+    match generate_metrics_data(&workspace, &BENCHMARK_PRIMES[0], 500, 500) {
+        Err(_) => eprintln!("Error"),
+        _ => {}
+    }
+
+    // Check whether the statement is true.
+    let mut simulator = Simulator::default();
```
[the remark this answers is not in the document]
Comment (Michael-Qedit) [reply rust/src/producers/circuit_generator.rs:253]: evaluator is for the sieve_ir repo. On zkinterface, it's still called simulator

File: rust/src/producers/circuit_generator.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    let ws: Workspace = Workspace::from_dir(&workspace.join(format!("metrics_{}_500_500/", &BENCHMARK_PRIMES[0])) ).unwrap();
+
+    // Must validate and simulate in parallel to support stdin.
+    for msg in ws.iter_messages() {
+        simulator.ingest_message(&msg);
+    }
+
+    assert_eq!(simulator.get_violations().len(), 0);
+
+}
```
[the remark this answers is not in the document]
Comment (Michael-Qedit) [reply rust/src/producers/circuit_generator.rs:263]: Done

File: rust/src/producers/circuit_generator.rs (hunk 1 above)
Comment (Michael-Qedit) [reply rust/src/producers/circuit_generator.rs:226]: I finally chose this even cleaner solution:
`
witnesses.iter().zip(bit_vector.iter()).map(|(idx, buint)| buint * idx ).sum::<BigUint>()
`
which ensures that only slices of same length will be used (avoiding reading out of bounds).

File: rust/src/producers/circuit_generator.rs [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            ((wit_idx.clone(), b1), (wit_idx.clone(), b2), (vec![instance_id], vec![1]))
+        ];
+        builder.push_constraints(ConstraintSystem::from(new_constraint.as_ref()))?;
+
+    }
+
+    // if it remains constraints not yet pushed, push them.
+    if constraints_vec.len() > 0 {
+        builder.push_constraints(ConstraintSystem::from(constraints_vec.as_ref()))?;
+    }
```
Comment (PaulLaux) [inline rust/src/producers/circuit_generator.rs:177]: Now that you pushing every iteration, you don't need that check.

**Review by PaulLaux — APPROVED** [review]
Looks great, approved with one comment.

**Review by PaulLaux — APPROVED** [review]
looks good!

**Pull request merged**